impl/vfat/cogent/src/super.cogent

```diff
@@ -0,0 +1,725 @@
+    in res
+    | Success () ->
+      let bh = brelse_ac bh
+      and dir = incr_version dir
+      in (nr_slots == 0)
+      | True ->
+        let dir = sync_inode_cog dir
+        in ((ex, dir), Success ())
+      | False ->
+        let slot_off = sinfo.slot_off
+        in deep_fat_remove_entries_ac (dir, slot_off, nr_slots)
+        | Success dir ->
+          let dir = sync_inode_cog dir
+          in ((ex, dir), Success ())
+        | Error dir ->
+          let sb = get_sb dir ! dir
+          and ex = fat_msg_ac (ex, sb, "Couldn't remove the long name slots")
+          and dir = sync_inode_cog dir
+          in ((ex, dir), Success ())
+    | Error err_ptr ->
+      let bh = brelse_ac (bh)
+      in ((ex, dir), Error err_ptr)
+  | False ->
+    let bh = brelse_ac(bh)
+    and dir = incr_version (dir)
+    in (nr_slots == 0)
+    | True ->
+      let dir = sync_inode_cog dir
+      in ((ex, dir), Success ())
+    | False ->
+      let slot_off = sinfo.slot_off
+      in deep_fat_remove_entries_ac (dir, slot_off, nr_slots)
+      | Success dir ->
+        let dir = sync_inode_cog dir
+        in ((ex, dir), Success ())
+      | Error dir ->
+        let sb = get_sb dir ! dir
+        and ex = fat_msg_ac (ex, sb, "Couldn't remove the long name slots")
+        and dir = sync_inode_cog dir
+        in ((ex, dir), Success ())
+
+fat_hash_cog: U64 -> U64
+fat_hash_cog ipos =
+  hash_32_ac [U64] (ipos)
+
+-- fairly sure this only reads but need to check what MSDOS_I does for sure
+fat_dir_hash_cog: VfsInode -> (VfsInode, U64)
+fat_dir_hash_cog inode =
+    let (inode, logstart) = get_logstart inode
+    in (inode, hash_32_ac [U32] logstart)
+
+fat_detach_cog: VfsInode -> VfsInode
+fat_detach_cog inode =
+  let sb = get_sb inode ! inode
+  and sb = spinode_lock sb
+  and fat_hash = get_fat_hash inode ! inode
+  and inode = set_ipos (inode, 0)
+  and _ = hlist_del_init_ac fat_hash
+  and _ = spinode_unlock (sb)
+  in is_dir_ac (inode) ! inode
+  | True ->
+    is_nfs inode ! inode
+    | True ->
+      let sb = get_sb inode ! inode
+      and sb = spdir_lock sb
+      and dir_hash = get_dir_hash inode ! inode
+      and _ = hlist_del_init_ac dir_hash
+      and _ = spdir_unlock sb
+      in inode
+    | False -> inode
+  | False -> inode
+
+-- handling overflow - need some error handling here
+incr_version: VfsInode -> VfsInode
+incr_version inode =
+  let version = get_version inode ! inode
+  and version = version + 1
+  in set_version(inode, version)
+
+setup_inode: (VfsInode, VfsDentry, Time) -> (VfsInode, VfsDentry)
+setup_inode (inode, dentry, ts) = -- check if order of when increasing i_version matters
+  let inode = incr_version inode
+  and inode = set_time (inode, ts)
+  and (inode, dentry) = d_instantiate_ac (inode, dentry)
+  in (inode, dentry)
+
+fat_build_inode_cog: (Superblock, SlotInfo) -> RR Superblock VfsInode ErrPtr
+fat_build_inode_cog (sb, sinfo) =
+  let bh = sinfo.bh
+  and bh = brelse_ac bh
```

Comment:
name shadowing. could be composed because the first `bh` binder is no longer visible.
just do `let bh = brelse_ac (sinfo.bh)`
also line 355 with `inode`